src/display/canvas.js

```diff
@@ -1082,11 +1082,19 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
       var ctx = this.ctx;
       var current = this.current;
       var x = current.x, y = current.y;
+      var xw, yh, origX, origY, pixelSize = this.getSinglePixelWidth();
       for (var i = 0, j = 0, ii = ops.length; i < ii; i++) {
         switch (ops[i] | 0) {
           case OPS.rectangle:
-            x = args[j++];
-            y = args[j++];
+            origX = args[0];
+            origY = args[1];
+            if (pixelSize) {
+              x = Math.floor(args[j++] / pixelSize) * pixelSize;
```

Comment:
I am new to html canvas's but here was my logic. Please verify I am looking at this correctly.
The goal is to extend all images outward to the nearest pixel. Compressing inward would generate gaps between images.

With each (x,y) point represented in the canvas with (0,0) being the top left most point.

Here, we are drawing a rectangle. So the goal is to extend each point as follows:

Top Left (x,y):  x rounds towards left (floor). Y rounds north towards y=0 (ceil)
Top Right (xw, y): xw rounds towards right (ceil). Y rounds north (ceil).
Bottom Right (xw, yh): xw rounds towards right (ceil). yh rounds south (while represented by positive value(?), so, to make it as large as possible, ceil.
Bottom Left (x, yh): x rounds towards left (floor). yh rounds south (ceil).

I have some confusion over how the Y variable is treated here. Is its value negative, considering its under the y=0 line (top of document)? Var height is set to negative in my example cases, while var Y is set to positive. Var YH is defined as height+y, where Y is normally always larger than height. This confusion could possibly lead to incorrect rounding.

Reply:
@timvandermeij  If you have a moment, please review this comment. I would like a second opinion before we merge, it may be possible that I have made a mistake in my decision to use .ceil / .floor in specific cases.

Reply:
Actually, it seems I get better results with the following:
```
x = Math.floor(origX / pixelSize) * pixelSize;
y = Math.floor(origY / pixelSize) * pixelSize;

xw = Math.ceil((origX + width) / pixelSize) * pixelSize;
yh = Math.ceil((origY + height) / pixelSize) * pixelSize;
```
I believe that the canvas actually treats Y height as positive, with y=0 being the top of the page and y>0+n = bottom.

Where my confusion comes from, I think, is the example file that I am using. When drawing rectangles from the top of the page to the bottom, the following values are being passed into "constructPath": (571.41, 684.211, 5.303, -578.104)

These values are X, Y, Width, Height respectively. What is strange here is that, according to my believe of X/Y values on a canvas, (X,Y) actually represents the lower left hand corner of our rectangle instead of the expected upper left corner in this example. 

When YH is calculated, we are adding the negative height to our Y value, giving us the top side of our rectangle. In this case, we would want to round this value towards the top of the page (y=0) so we would want to floor this value.

However, if the usual case is that (X,Y) is the upper left corner of our image (like I originally expected), that would mean that height would be a positive value, and therefore YH would represent the bottom of our rectangle. In this case, we would want to make YH rounded upward towards the bottom of the document, aka Math.ceil(YH).

I am not too sure what determines the values of (X,Y, Height, Width) when passed into "constructPath." Can an assumption ever be made about which point (top left, top right, bottom left, bottom right) (X,Y) actually represents? If not, I need to add more code to determine the location of (X,Y) before making assumptions about the direction of rounding.

Reply:
@brendandahl